google-cloud-spanner/src/main/java/com/google/cloud/spanner/SimpleClientImpl.java

```diff
@@ -0,0 +1,139 @@
+      Statement statement;
+      if (args != null && !args.isEmpty()) {
+        Statement.Builder builder = Statement.newBuilder(query);
+        for (Map.Entry<String, Value> entry : args.entrySet()) {
+          builder.bind(entry.getKey()).to(entry.getValue());
+        }
+        statement = builder.build();
+      } else {
+        statement = Statement.of(query);
+      }
+      return tx.executeQuery(statement);
+    }
+  }
+
+  @Override
+  public ResultSet executeSqlQuery(String query, Map<String, Value> args) {
+    return executeSqlQuery(query, args, TimestampBound.strong());
+  }
+
+  @Override
+  public ResultSet executeSnapshotSqlQuery(String query, Map<String, Value> args) {
+    return executeSqlQuery(query, args, TimestampBound.ofExactStaleness(30L, TimeUnit.SECONDS));
+  }
+
+  @Override
+  public long executeSqlWrite(String sql, Map<String, Value> args) {
+    final Statement.Builder builder = Statement.newBuilder(sql);
+    if (args != null && !args.isEmpty()) {
+      for (Map.Entry<String, Value> entry : args.entrySet()) {
+        builder.bind(entry.getKey()).to(entry.getValue());
+      }
+    }
+    return client
+        .readWriteTransaction()
+        .run(
+            new TransactionRunner.TransactionCallable<Long>() {
+              @Nullable
+              @Override
+              public Long run(TransactionContext transaction) throws Exception {
+                return transaction.executeUpdate(builder.build());
+              }
+            });
+  }
+
+  @Override
+  public boolean runTransaction(final List<Statement> statements) {
+    try {
+      return client
+          .readWriteTransaction()
+          .run(
+              new TransactionRunner.TransactionCallable<Boolean>() {
+                @Override
+                public Boolean run(TransactionContext transaction) {
+                  try {
+                    transaction.batchUpdate(statements);
+                    return true;
+                  } catch (Exception e) {
+                    logger.log(Level.WARNING, "Failed to complete transaction", e);
+                  }
+                  return false;
+                }
+              });
+    } catch (Exception e) {
+      throw SpannerExceptionFactory.newSpannerException(e);
+    }
+  }
+
+  @Override
+  public void runTransaction(final Statement readStatement, final OnReadHandler handler) {
+    client
+        .readWriteTransaction()
+        .run(
+            new TransactionRunner.TransactionCallable<Boolean>() {
+              @Override
+              public Boolean run(TransactionContext transaction) throws Exception {
+                try {
+                  ResultSet resultSet = transaction.executeQuery(readStatement);
+                  handler.handle(resultSet, transaction);
+                  return true;
+                } catch (Exception e) {
+                  logger.log(Level.WARNING, "Failed to complete transaction", e);
```

Comment:
The Spanner client library will commit the transaction if the `TransactionCallable#run(TransactionContext)` method returns normally. As all exceptions are caught by this catch clause, the transaction will always be committed, regardless of the type of exception that might occur during the transaction.

It will also prevent automatic retries of any aborted transaction, as that is based on the `AbortedException` bubbling up from the `run` method.

Catching any exception here will cause the transaction to commit instead of rollback, as the `run` method returns normally. The client library will

Reply:
change to void
Now not catching exceptions.